velox/functions/prestosql/Sequence.cpp

```diff
@@ -0,0 +1,173 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include "velox/expression/VectorFunction.h"
+#include "velox/vector/ConstantVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+class SequenceFunction : public exec::VectorFunction {
+ public:
+  static constexpr int32_t kMaxResultEntries = 10'000;
+
+ private:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VectorPtr localResult;
+    if (args[0]->type() == BIGINT()) {
+      localResult = applyNumber(rows, args, outputType, context);
+    }
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+  VectorPtr applyNumber(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    const auto numRows = rows.size();
+    auto pool = context.pool();
+    int totalCount = 0;
+
+    auto startVector = args[0];
+    auto stopVector = args[1];
+    VectorPtr stepVector;
+    if (args.size() == 3) {
+      stepVector = args[2];
+    }
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      int count = sequenceCount(startVector, stopVector, stepVector, row);
+      totalCount += count;
+    });
+
+    VectorPtr elements = BaseVector::create(BIGINT(), totalCount, pool);
+    BufferPtr sizes = allocateSizes(numRows, pool);
+    BufferPtr offsets = allocateOffsets(numRows, pool);
+    auto rawElements = elements->asFlatVector<int64_t>()->mutableRawValues();
+    auto rawSizes = sizes->asMutable<vector_size_t>();
+    auto rawOffsets = offsets->asMutable<vector_size_t>();
+
+    int elementsOffset = 0;
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      rawOffsets[row] = elementsOffset;
+      rawSizes[row] = sequenceCount(startVector, stopVector, stepVector, row);
+      writeToElements(
+          rawElements,
+          elementsOffset,
+          startVector,
+          stopVector,
+          stepVector,
+          row);
+      elementsOffset += rawSizes[row];
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool, outputType, nullptr, numRows, offsets, sizes, elements);
+  }
+
+  int sequenceCount(
+      VectorPtr startVector,
+      VectorPtr stopVector,
+      VectorPtr stepVector,
```

Comment:
VectorPtr is:
```
using VectorPtr = std::shared_ptr<BaseVector>;
```

Passing it by value will incur ref increase and then decrease, which would be a performance hit, especiall done per row.
Consider passing by const ref:
```
const VectorPtr& startVector,
...
```